src/app/shared/context.service.ts

```diff
@@ -113,75 +114,41 @@ export class ContextService implements Contexts {
-        })
-        .filter(([changedContext, recentContexts, index]: [Context, Context[], number]): boolean => {
-          return index !== 0; // continue only if the changed context is new, or requires a move in _recent
-        })
-        .map(([changedContext, recentContexts, index]: [Context, Context[], number]): Context[] => {
-          // if changedContext exists in _recent
-          if (index !== -1) {
-            // move the context to the front of the list if it is already in _recent
-            if (recentContexts[0].name !== changedContext.name) {
-              recentContexts.splice(index, 1);
-              recentContexts.unshift(changedContext);
-            }
-            // otherwise, no operation required; changedContext is already index 0
-          // if changedContext is new to the recent contexts
-          } else {
-            recentContexts.unshift(changedContext);
-            // trim _recent if required to ensure it is length =< 8
-            if (recentContexts.length > this.RECENT_CONTEXT_LENGTH) {
-              recentContexts.pop();
-            }
-          }
-          return recentContexts;
-        })
-        .subscribe((recentContexts: Context[]): void => {
-          this.saveRecent(recentContexts);
+        .pipe(
+          withLatestFrom(this.recent),
+          map(([changedContext, recentContexts]: [Context, Context[]]): RecentData<Context> => {
+            return this.onBroadcastChanged(changedContext, recentContexts);
+          }),
+          filter((recentData: RecentData<Context>): boolean => recentData.isSaveRequired === true)
+          )
+        .subscribe((recentData: RecentData<Context>): void => {
+          this.saveRecentContexts(recentData.data);
         })
     );
 
     this.subscriptions.push(
       this.broadcaster.on<Space>('spaceDeleted')
-      .withLatestFrom(this.recent)
-      .map(([deletedSpace, recentContexts]: [Space, Context[]]): [Space, Context[], number] => {
-        let index: number = recentContexts.findIndex((context: Context): boolean => {
-          if (context.space) {
-            return context.space.id === deletedSpace.id;
-          } else {
-            return false;
-          }
-        });
-        return [deletedSpace, recentContexts, index];
-      })
-      .filter(([deletedSpace, recentContexts, index]: [Space, Context[], number]): boolean => {
-        return index !== -1;
-      })
-      .map(([deletedSpace, recentContexts, index]: [Space, Context[], number]): Context[] => {
-        recentContexts.splice(index, 1);
-        return recentContexts;
-      })
-      .subscribe((recentContexts: Context[]): void => {
-        this.saveRecent(recentContexts);
-      })
+        .pipe(
+          withLatestFrom(this._recent),
+          map(([deletedSpace, recentContexts]: [Space, Context[]]): RecentData<Context> => {
+            return this.onBroadcastDeleted(deletedSpace, recentContexts);
```

Comment:
Just a style thing but I would drop the curly braces and `return` here, personally.